### PR TITLE
MPAS-Ocean threading and GPU update

### LINKED ISSUE #2860: Errors from /opt/cray/gni-headers "dla_retransmit_discarded_blocks:" for several threaded MPAS tests on cori-knl
New failures that look to me like system-level fails (ie unrelated to our code, but I've not verified). I found 2 of these a few days ago after the Cori SW upgrade (April 10) and was trying to find out more. I had not realized that one of our routine tests were also failing with what looks like a similar error -- but has been happening since March 30th. There _could_ have been some changes on the system in preparation for the Cori SW upgrade. Or it could be something else entirely.

```
With 
SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6
SMS_D_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6

3885: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari/include/gni_dla_state.h:1574: dla_retransmit_discarded_blocks: Assertion `dla_hndl->first_assigned_blk == first_discarded_block' failed.

And here is another example of errors from this run. It's not always clear which error caused the original problem:

   0:  WARNING: Rearr optional argument is a pio2 feature, ignored in pio1
3421: Fatal Error: unknown request type: 0 at line 1163 in file cqe_processing.c
3601: Assertion failed in file /notbackedup/users/sko/mpt/mpt_base/mpich2/src/mpid/ch3/channels/nemesis/include/mpid_nem_inline.h at line 878: el->usage
3530: Assertion failed in file /notbackedup/users/sko/mpt/mpt_base/mpich2/src/mpid/ch3/src/ch3u_request.c at line 692: MPIU_Object_get_ref((req)) >= 0
3314: Fatal Error: unknown request type: 0 at line 1163 in file cqe_processing.c
3075: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
3075: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari/include/gni_dla_state.h:1461: dla_retransmit_request: Assertion `num_blocks <= dropped_req->num_dla_blocks' failed.
3074: forrtl: severe (174): SIGSEGV, segmentation fault occurred
3309: Assertion failed in file /notbackedup/users/sko/mpt/mpt_base/mpich2/src/mpid/ch3/channels/nemesis/include/mpid_nem_inline.h at line 878: el->usage
3079: forrtl: severe (174): SIGSEGV, segmentation fault occurred
3069: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.

```

```
With 
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel
ERS_Ld5.T62_oQU120.CMPASO-NYF.cori-knl_intel

 56: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
 56: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari/include/gni_fma.h:480: GNII_GenAllocSeqid: Assertion `((dla_hndl)->blocks[(local_block_id)].state == 1)' failed.
```
For the above two fails, I sent email to NERSC ( INC0136942 ). I suggested it could be related to the SW upgraded. NERSC suggested I try with updated modules. I have tried a few different module changes with no effect yet.

Using the now saved test directories, I can see:
```
cori11% grep -m1 "dla_retransmit_" /global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd2019*/e3sm.log.*
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190330_004404/e3sm.log.20122776.190330-015942:3079: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1461: dla_retransmit_request: Assertion `num_blocks <= dropped_req->num_dla_blocks' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190330_193029/e3sm.log.20138829.190330-201537:3073: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1461: dla_retransmit_request: Assertion `num_blocks <= dropped_req->num_dla_blocks' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190331_193045/e3sm.log.20166482.190331-224316:3079: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1530: dla_retransmit_discarded_blocks: Assertion `((((unsigned int)(first_discarded_block)) < ((dla_hndl)->num_blocks)) ? 1 : 0)' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190401_221505/e3sm.log.20181706.190402-001418:3283: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1461: dla_retransmit_request: Assertion `num_blocks <= dropped_req->num_dla_blocks' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190402_193027/e3sm.log.20199017.190402-202124:3079: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1461: dla_retransmit_request: Assertion `num_blocks <= dropped_req->num_dla_blocks' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190403_193049/e3sm.log.20234597.190403-203254:3079: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190404_193146/e3sm.log.20255561.190404-201301:3079: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1461: dla_retransmit_request: Assertion `num_blocks <= dropped_req->num_dla_blocks' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190405_193046/e3sm.log.20279367.190405-201134:3079: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190406_193132/e3sm.log.20284843.190407-150251:3788: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190407_193057/e3sm.log.20296450.190407-204409:3954: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190408_193114/e3sm.log.20337970.190408-201713:3079: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190409_193100/e3sm.log.20374458.190409-201520:3069: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190410_204838/e3sm.log.20386710.190411-105323:3883: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190411_193141/e3sm.log.20405021.190411-202336:3962: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190412_193041/e3sm.log.20439775.190412-211827:3282: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190413_193048/e3sm.log.20461801.190413-202533:3947: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6.C.JNextProd20190414_193134/e3sm.log.20499896.190414-202052:3885: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari/include/gni_dla_state.h:1574: dla_retransmit_discarded_blocks: Assertion `dla_hndl->first_assigned_blk == first_discarded_block' failed.
```

Can also see that the ERS_ test (using only 2 nodes) also started failing on March 31st: 
```
cori01% grep -m1 dla /global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/ERS_Ld5.T62_oQU120.CMPASO-NYF.cori-knl_intel.JNextDeveloper20190*/e3sm*
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/ERS_Ld5.T62_oQU120.CMPASO-NYF.cori-knl_intel.JNextDeveloper20190331_090552/e3sm.log.20154532.190331-094652: 59: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/ERS_Ld5.T62_oQU120.CMPASO-NYF.cori-knl_intel.JNextDeveloper20190401_090717/e3sm.log.20170067.190401-095959: 56: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/ERS_Ld5.T62_oQU120.CMPASO-NYF.cori-knl_intel.JNextDeveloper20190403_090558/e3sm.log.20220300.190403-101235: 59: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/ERS_Ld5.T62_oQU120.CMPASO-NYF.cori-knl_intel.JNextDeveloper20190407_151236/e3sm.log.20292497.190407-155810: 59: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/ERS_Ld5.T62_oQU120.CMPASO-NYF.cori-knl_intel.JNextDeveloper20190409_090805/e3sm.log.20358555.190409-095007: 59: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.0_24.1__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/ERS_Ld5.T62_oQU120.CMPASO-NYF.cori-knl_intel.JNextDeveloper20190411_111354/e3sm.log.20397496.190411-115204:101: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
/global/cscratch1/sd/acmetest/acme_scratch/cori-knl/old_test_archive/old_runs/ERS_Ld5.T62_oQU120.CMPASO-NYF.cori-knl_intel.JNextDeveloper20190413_090701/e3sm.log.20446582.190413-095616: 56: e3sm.exe: /opt/cray/gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari/include/gni_dla_state.h:1567: dla_retransmit_discarded_blocks: Assertion `rc == GNI_RC_SUCCESS' failed.
```

And looking in all runs, I see that these 2 other tests are impacted:
```
SMS.T62_oRRS30to10v3wLI.GMPAS-IAF.cori-knl_intel
SMS.ne120_oRRS18v3_ICG.A_WCYCL2000_H01AS.cori-knl_intel.cam-cosplite
```

### PR DESCRIPTION
**Pull request by mark-petersen**
This updates OpenMP and OpenACC calls in MPAS-Ocean, including:
- GPU optimization in EOS: https://github.com/MPAS-Dev/MPAS-Model/pull/69 and https://github.com/MPAS-Dev/MPAS-Model/pull/170
- Update ocean BGC threading: https://github.com/MPAS-Dev/MPAS-Model/pull/152
- Reduce threading barriers: https://github.com/MPAS-Dev/MPAS-Model/pull/194

It is not bfb due to change in order of operations. It should not be climate changing.
Fixes #2860

[non-BFB]

### CODE REVIEW

**Comment by mark-petersen** [comment]
The only addition here is 5292750. I had to put it on top of #2843 for everything to work with the new Makefiles. Once #2843 is merged, it will only show that last commit, which is all the OpenMP and OpenACC change.

**Comment by mark-petersen** [comment]
@philipwjones and @amametjanov please look over the merged commits in MPAS-Ocean, particularly the file:
`
src/core_ocean/shared/mpas_ocn_equation_of_state_jm.F
`
which had several conflicts between the first and later MPAS PRs. I resolved those by hand by removing  the private clause on the omp lines, but please check that.

**Comment by mark-petersen** [comment]
@jgfouca please test this on your standard machines. It includes an EOS GPU update that we merged in and then reverted due to a threading problem that only came up on one of your machines.  That is fixed on the second commit here: https://github.com/MPAS-Dev/MPAS-Model/pull/170/commits

**Comment by jgfouca** [comment]
@mark-petersen , this branch does not build for me on melvin:

```
./create_test SMS_P12x2.ne4_oQU240.A_WCYCL1850.melvin_gnu --no-run
...
ERROR: BUILD FAIL: mpascice.buildlib failed
```

Error is:
```
f951: Warning: Nonexistent include directory ‘/home/jgfouca/acme/scratch/SMS_P12x2.ne4_oQU240.A_WCYCL1850.melvin_gnu.20190412_120048_osduo3/bld/ice/source/inc’ [-Wmissing-include-dirs]

ice_comp_mct.f90:869:63:

         call seaice_timestep(domain, domain % clock, itimestep)
                                                               1
Error: Missing actual argument for argument ‘ierr’ at (1)
```

**Comment by mark-petersen** [comment]
@jgfouca Please try again, I made a mistake in merging the submodule. I got the same error on my end but it works now.  Please run tests that exercise threading, like PET tests with various compilers.

**Comment by mark-petersen** [comment]
Passes the following. Note it includes MPAS tests using debug on gnu and intel, and PET threading tests with both gnu and intel. 
```
PASS SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu RUN time=810
PASS SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel RUN time=383
PASS ERS_Ld9.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_gnu COMPARE_base_rest
PASS PEM_Ln3.T62_oEC60to30v3wLI.GMPAS-IAF-ISMF.cori-knl_gnu COMPARE_base_modpes
PASS PEM_Ln9.ne30_oECv3wLI.A_WCYCL1850-DIB-ISMF_CMIP6.cori-knl_gnu COMPARE_base_modpes
PASS PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_gnu COMPARE_base_single_thread
PASS PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_intel COMPARE_base_single_thread
PASS PET_Ln3.ne30_oECv3wLI.A_WCYCL1850-DIB-ISMF_CMIP6.cori-knl_gnu COMPARE_base_single_thread
```

**Comment by jgfouca** [comment]
@mark-petersen looks good now, thanks.

**Comment by jgfouca** [comment]
Can I start integrating this?

**Comment by mark-petersen** [comment]
The Makefile changes #2843 needs to be merged in first.  This branch was put on top of that one because the Makefile changes need to be done in MPAS and E3SM repos together.

File: components/mpas-albany-landice/cime_config/buildlib [hunk 1]
```diff
@@ -6,7 +6,7 @@ my $component = "mpas-albany-landice";
 my $component_dir = "glc";
 my $component_driver = "glc_cesm_driver";
 my $core = "landice";
-my $esm = "ACME";
+my $esm = "E3SM";
```
Comment (jonbob) [inline components/mpas-albany-landice/cime_config/buildlib]: @mark-petersen - these weren't in the previous PR?

Comment (mark-petersen) [reply components/mpas-albany-landice/cime_config/buildlib]: I have this pointing to a temporary branch e3sm/testing, so it's still showing previous diffs.  I'll rebase...

Comment (mark-petersen) [reply components/mpas-albany-landice/cime_config/buildlib]: Done.

**Comment by mark-petersen** [comment]
@amametjanov or @philipwjones could one of you do your standard threaded test on this PR, and measure any before/after performance difference? I'm guessing this work is more clean-up and GPU prep, but it is good to document any threaded performance change.

@jonbob This diffs from master, bd399bd2e only by this one submodule update.  Could you run two months of low-resolution B case with and without this commit, so we can compare changes exactly?

**Comment by mark-petersen** [comment]
Rebased and tested again:
```
PASS SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu RUN time=661
PASS SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel RUN time=353
PASS PEM_Ln9.ne30_oECv3wLI.A_WCYCL1850-DIB-ISMF_CMIP6.cori-knl_gnu COMPARE_base_modpes
PASS PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_gnu COMPARE_base_single_thread
```

**Comment by philipwjones** [comment]
@mark-petersen - ran a small G case in debug queue, but did not see any noticeable difference in timing between this and the latest master

**Comment by amametjanov** [comment]
Yes, this is a clean-up and bug-fix PR to enable threading in BGC runs: threaded runs should be BFB reproducible after this.

**Comment by ndkeen** [comment]
Would like to see this in. Fixes #2860

**Comment by mark-petersen** [comment]
All this code is now merged into MPAS-Dev repository. Rebased and retested. This is ready to merge into next. Note it is non-bit-for-bit.

**Comment by mark-petersen** [comment]
Testing so far, after yesterday's rebase:
```
PASS SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu RUN time=189
PASS SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel RUN time=388
PASS PEM_Ln3.T62_oEC60to30v3.GMPAS-IAF.cori-knl_gnu COMPARE_base_modpes
PASS PET_Ln3.T62_oEC60to30v3.GMPAS-IAF.cori-knl_gnu COMPARE_base_single_thread
PASS PET_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_gnu COMPARE_base_single_thread
```
I'm getting failed runs on domains with ice shelf cavities. Looking into it now.

**Comment by mark-petersen** [comment]
I have a test that fails on both master and this branch, so is not relevant to this PR:
```
./create_test SMS_D_Ln9.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_gnu -p m2833 -q debug --walltime 00:30:00
```
fails on initialization with:
```
cd /global/cscratch1/sd/mpeterse/acme_scratch/cori-knl/SMS_D_Ln9.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_gnu.20190528_063017_qt5v06/run
cori02:run$ tail -n 1 cpl.log.21795489.190528-065238
 ERROR: (seq_mct_drv) ERROR: if glc_c2_ocn must also have ocn_c2_glc and vice versa. Boundary layer fluxes calculated in coupler require input from both components.
```
I was not able to conduct a bfb thread test with melt fluxes on after the rebase. However, before yesterday's rebase I did conduct that test and it passed PET (this is what @matthewhoffman asked this morning), so I think we are safe here. See https://github.com/E3SM-Project/E3SM/pull/2852#issuecomment-487092566

**Comment by matthewhoffman** [comment]
@mark-petersen , what I was suggesting is that you could test merge your branch into ef9e1b9f3, which is the commit prior to the branch that Jon and I introduced last week that leads to the error you are seeing.  That way you would be able to get past that error and do further testing.  Alternatively, you could wait a couple days and we will have a fix for that error in place.

**Comment by mark-petersen** [comment]
Rebased with today's ISMF fix. Retested with ocean cavities, everything looks fine:
```
PASS SMS_D_Ln9.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_gnu RUN time=262
PASS SMS_D_Ln9.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_intel RUN time=1101
PASS PEM_Ln9.ne30_oECv3wLI.A_WCYCL1850-DIB-ISMF_CMIP6.cori-knl_gnu COMPARE_base_modpes
PASS PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_gnu COMPARE_base_single_thread
```
This is ready to merge into next when there is space for a non-bfb PR.

**Comment by jonbob** [comment]
merged to next

**Comment by jonbob** [comment]
merged to master

**Pull request merged**